**2. What is the benefit of lambda?**

The benefit of using lambda functions, also known as anonymous functions, is that they can be defined in a concise and readable way, allowing for short and simple expressions to be passed as arguments or assigned to variables. This can make code more readable and maintainable. Lambda functions are also useful when a small piece of functionality needs to be passed as an argument to a higher-order function, such as map(), filter(), or reduce(), without having to define a separate function.

Additionally, Lambda functions are generally more lightweight and have less overhead than defining a function with a name, as they are generally used for a single time.

In Python programming language you can use it for example for filter to select only specific elements from an iterable and map to manipulate each element of an iterable.

**3. Compare and contrast map, filter, and reduce.**

map(), filter(), and reduce() are all higher-order functions that take a function and an iterable as arguments and return a new iterable.

map() *applies a given function to each element* of an iterable, and returns a new iterable containing the modified elements. It takes two arguments: the first is a function that takes one argument, and the second is an iterable. The function passed to map() is applied to each element of the iterable, and the result is a new iterable containing the modified elements.

In [5]:
numbers = [1, 2, 3, 4, 5]
squares = map(lambda x: x**2, numbers)
print(list(squares))

words = ['cat', 'dog', 'bird']
lengths = map(len, words)
print(list(lengths))

[1, 4, 9, 16, 25]
[3, 3, 4]


filter() returns an iterable containing only the elements that satisfy a given condition. It takes two arguments: the first is a function that takes one argument and returns a Boolean value, and the second is an iterable. The function passed to filter() is applied to each element of the iterable, and the result is a new iterable containing only the elements for which the function returned True.

In [4]:
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
evens = filter(lambda x: x % 2 == 0, numbers)
print(list(evens))

words = ['cat', 'dog', 'bird', 'fish', 'monkey']
long_words = filter(lambda x: len(x) > 3, words)
print(list(long_words))

[2, 4, 6, 8, 10]
['bird', 'fish', 'monkey']


reduce() is an implementation of a mathematical concept called folding and it applies a binary function (accepting two elements) cumulatively on the elements of an iterable, from left to right, so as to reduce the iterable to a single value. It takes two arguments: a function that takes two arguments and returns a single value, and an iterable. The function passed to reduce() is applied to the first two elements of the iterable, then to the result and the next element, and so on, until the iterable is reduced to a single value.
Keep in mind that reduce() is not available in Python built-in function and you have to import it from functools library.

In [6]:
from functools import reduce

numbers = [1, 2, 3, 4, 5]
product = reduce(lambda x, y: x*y, numbers)
print(product)

words = ['cat', 'dog', 'bird']
concatenation = reduce(lambda x, y: x + y, words)
print(concatenation)

120
catdogbird


**4. What are function annotations, and how are they used?**


Function annotations are a way to provide a programmer's intended type for a function's arguments and return value. They are not enforced by the Python interpreter and are used mainly for documentation purposes, but they can be used by third-party type checkers and IDEs. They are written in the form of name: type after the function argument, and they can also be used to specify the return type using -> type after the function arguments.

**5. What are recursive functions, and how are they used?**

A recursive function is a function that calls itself. It has a base case that stops the recursion, and one or more recursive cases that call the function again with different arguments. This allows the function to solve a problem by breaking it down into smaller instances of the same problem. Recursive functions are used for solving problems that have a natural recursive structure, such as traversing a tree data structure or computing factorials.





In [7]:
def factorial(n: int) -> int:
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)


In this example, the base case is when n is equal to 0, in which case the function returns 1. In the recursive case, the function calls itself with the argument n-1 and multiplies the result by n. This will keep happening until n reaches 0, it stops and return the final result.

For example, if we call factorial(5), it will evaluate as:

5 * factorial(4) -> 5 * 4 * factorial(3) -> 5 * 4 * 3 * factorial(2) -> 5 * 4 * 3 * 2 * factorial(1) -> 5 * 4 * 3 * 2 * 1 * factorial(0) -> 5 * 4 * 3 * 2 * 1 * 1 = 120

This is the mathematical definition of factorial, which is the product of all the integers from 1 to n (5!), and the recursive implementation follows that concept.

**6. What are some general design guidelines for coding functions?**


Some general design guidelines for coding functions are:
Keep functions small and focused. A function should have a single, well-defined purpose and should not try to do too many things.

Give functions descriptive and meaningful names. The name of a function should clearly indicate what it does.

Use function arguments to make the function flexible and reusable. Instead of hardcoding values in a function, use arguments to allow the caller to specify them.

Use clear and descriptive variable names. The variable names used within a function should be clear and meaningful, making it easy to understand the function's purpose.

Use meaningful return values. A function's return value should be used to communicate its results to the caller.

**7. Name three or more ways that functions can communicate results to a caller.**

Some ways that functions can communicate results to a caller are:

Return statement: Functions can use the return statement to return a value or expression to the caller. For example, a function that calculates the factorial of a number can return the result as its return value.

Side effect: A function can also communicate results to a caller by changing the state of one or more objects outside of the function, known as side effects.

Exception: Functions can raise an exception to signal to the caller that an error has occurred. This can be used to indicate an exceptional condition that the caller must handle.

Output parameters : Functions can take input and output parameters, in other words, passed by reference. A function can change the value of the parameter, this change would reflect back in the calling function.